### ISSUE TITLE
Add support for "Icon Packs" 

### ISSUE THREAD

**Issue by AObuchow** [eclipse-platform/eclipse.platform.ui#13]
This is a continuation/migration of [Bug 562174](https://bugs.eclipse.org/bugs/show_bug.cgi?id=562174) - Add support for "icon packs" in Eclipse IDE.

It'd be nice for to have an officially supported way to replace the icons used in Eclipse.

There is a lot of discussion to be held on how this should be handled (and there was already a lot of discussion on the original BZ thread), so I'm going to leave this relatively open-ended.

IIRC the approach I had initially tried was:
- Allow custom/replacement icons to be provided by a plugin fragment, by placing them in a "theme" directory of the plugin fragment
    - So in order to replace one of the "debug" icons, you must make a plugin fragment where the host plugin is "eclipse.debug.ui" and place the icon within a "theme" directory, with the same file path as the original icon
- When loading an icon from a plugin, check if the plugin provides an alternative path to the icon that is prefixed by "/themes/"
- If so, load it
- Otherwise, use the default icon

There are probably some limitations to this approach, for instance:
- The icon-provider must create a plugin fragment for each plugin which they wish to replace the icons for. This is very tedious.
- How do we handle switching between different icon packs, if each plugin fragment uses the same "themes" directory?

It might be nicer to set some JFace preference which points to a specific plugin path (where the plugin is the icon pack provider plugin), and then alternative icons are checked from that plugin before loading the default icon. 

I am going to submit a WIP PR that shows my current implementation, as well as an example plugin fragment that provides icon replacements.

**Comment by laeubi** [comment]
I think a preference would be a good idea in any case to prevent excessive lookups, even if they are fast they are not for free.

> t might be nicer to set some JFace preference which points to a specific plugin path

There is already `org.eclipse.e4.ui.css.swt.theme/themeid` I think it would be good to reuse that, lets assume the themeid is 'my.custom.theme' jface could simply prefix all icons with that id eg `icon/cross.png` -> `icon/my.custom.theme/cross.png` with a fallback to the original icon as the "error icon".

**Comment by mickaelistria** [comment]
> There is already org.eclipse.e4.ui.css.swt.theme/themeid

While it's good to reuse it if possible, I think it's important that this particular issue of "icon packs" can be built-in JFace without dependency of e4 or higher-level platform.
Then the integration of icon packs in e4 themes can decide to rely on this themeid, but that discussion is IMO a step further.

**Comment by laeubi** [comment]
> JFace without dependency of e4 or higher-level platform.

JFace don't need to depend on e4, just reading that preferences with that ID, I actually won't care much about that in contrast to inventing another "icon pack id" that then must match the themeId or similar ...

Actually it makes most sense to have a icon pack for a theme than having different icons with the standard theme.

**Comment by mickaelistria** [comment]
> Actually it makes most sense to have a icon pack for a theme than having different icons with the standard theme.

I disagree here. Also SWT/JFace do not apply only to Eclipse IDE nor workbench, there are other apps that may want icon/image theme without dragging the e4 stuff.
Let's fix the issue where it is (in JFace) without considering the legacy downstream (e4), that will lead to the best design.

**Comment by laeubi** [comment]
You don't need to drag in anything. This is just a preference and we can access it in any bundle we like without any dependency. As SWT does not allow for any theming, it might then even be better to *not* do this in swt/jface but in the eclipse theming framework to handle that case.

**Comment by BeckerWdf** [comment]
> * and place the icon within a "theme" directory, with the same file path as the original icon

So this implies only icons that originally come from a file can be replaced? What out other icons (that are e.g. drawn via GC operations) like these:
<img width="226" alt="image" src="https://user-images.githubusercontent.com/28338612/164488290-018ea030-d9d1-4a6e-8818-a47598205905.png">

**Comment by laeubi** [comment]
> What out other icons (that are e.g. drawn via GC operations) like these:

These are obviously not Image-Icons :-)

**Comment by BeckerWdf** [comment]
Will this new feature allow plugin developers to use different icons for the light and the dark theme? Currently it's quite hard to design icons that have good contrast on light and dark background. So providing alternative icons for the dark theme is a much needed feature.

> These are obviously not Image-Icons :-)

So if the answer to the above question is yes: We might also want to provide different versions of these "none-image-icons". What would you propose as as solution for this? Having IFs in the drawing code that check for theme-names or brightness of the background color?

**Comment by laeubi** [comment]
> Will this new feature allow plugin developers to use different icons for the light and the dark theme?

Not in its current form, thats why I requested cahnges and suggested to better integrate this with the theming framework instead of an additional soloution just for SWT.

> So if the answer to the above question is yes: We might also want to provide different versions of these "none-image-icons".

none-image-icons are custom painted components as as such are best supporting CSS styling instead of loading alternative images. see:
https://wiki.eclipse.org/E4/CSS/SWT_Mapping
https://wiki.eclipse.org/Eclipse4/RCP/CSS#Extending_CSS_to_handle_new_widgets

**Comment by BeckerWdf** [comment]
> Not in its current form, thats why I requested changes and suggested to better integrate this with the theming framework instead of an additional solution just for SWT.

👍

**Comment by mickaelistria** [comment]
> Will this new feature allow plugin developers to use different icons for the light and the dark theme?

Ultimately yes, but it will also require change in the theme support once we're done with SWT part. The theme would be able to define an icon set to use and configure SWT; so SWT would then load different images.

**Comment by becph01** [comment]
@AObuchow what would it take to get this supported for icon packs? I can look for budget to sponsor this.
It's for an RCP application to have a sane light mode/dark mode color scheme.

**Comment by laeubi** [comment]
See also here for some thoughts how to have "icon theming support":

- https://github.com/eclipse-platform/eclipse.platform.images/issues/81

**Comment by AObuchow** [comment]
> @AObuchow what would it take to get this supported for icon packs? I can look for budget to sponsor this. It's for an RCP application to have a sane light mode/dark mode color scheme.

@becph01 I actually think @laeubi would be a much better candidate to sponsor for developing this feature :)

@laeubi hopefully this is something you're allowed to do? I'm not sure if your work has some conflict-of-interest clause regarding sponsored development for the Eclipse IDE?

**Comment by laeubi** [comment]
> @laeubi hopefully this is something you're allowed to do? I'm not sure if your work has some conflict-of-interest clause regarding sponsored development for the Eclipse IDE?

Eclipse Foundation does not limit me regarding other agreements, especially if its in the best interest for the Eclipse IDE itself :-)

**Comment by laeubi** [comment]
FYI there is now an example for icon pack here with a description, please everyone try out and give feedback:
- https://github.com/eclipse-platform/eclipse.platform.ui/tree/master/examples/org.eclipse.ui.examples.iconpack

**Comment by HeikoKlare** [comment]
Thank you for providing the transformer mechanism and an example project, Christoph!
That sounds like a nice feature as it may be sufficient to address one of the use cases (deploy an RCP product with completely custom icons). I just wanted to test that out and found that the example works fine for me on Linux but not on Windows. I did the same on both systems:
- Take an SDK product configuration
- Add the transformer hooks fragment
- Add the iconpacks example bundle
- Add osgi framework extension argument

On Linux, the save icon is replaced while on Windows it is not. So far, I debugged down to `EquinoxBundle.getEntry(String)` where I see the difference that on Linux a `BundleFileWrapperChain` with the `TransformedBundleFile` is retrieved for a bundle on which a transformer was registered (like the `org.eclipse.ui` bundle in the example), while on Windows a simple `DirBundleFile` is retrieved.

Before digging deeper into it: @laeubi, are you aware of any OS-specifics in the functionality?

**Comment by laeubi** [comment]
@HeikoKlare I'm not aware of that, but I'm also only came across this recently, so maybe a bug no one noticed because it seems not widely used, it should of course work regardless of OS, but *might* get some problems due to `\` versus `/` things. At least that are exactly the things we only will find out by trying it 👍

**Comment by HeikoKlare** [comment]
Thanks for the quick reply! Alright, I will reinvestigate this. I just used another SDK in which I have org.eclipse.osgi in the workspace instead of the target platform and there it also works fine on Windows 👍 
With my usual workspace with org.eclipse.osgi coming from the target platform, basically all bundles are already initialized while the transformer hook is being setup (i.e., before the hook for creating transformed bundles is initialized), as the context of `org.eclipse.osgi` contains most if not all bundles of the run configuration (for whatever reason). It thus already initialized the `org.eclipse.ui` bundle before setting up the transformer hook:
![Image](https://github.com/user-attachments/assets/2da20414-ea96-4102-aec4-c03d3998914e)
Might be that there is just a weird configuratoin is place. But since I did nothing special but just extending an `sdk.product`, others might run into the same situation.

**Comment by HeikoKlare** [comment]
Some additional feedback and thanks to you, @laeubi 🏅:
I have just tested the mechanism in our RCP-based, Windows-only product to replace some of the Eclipse platform icons. It works as expected. So my first impression is that our use case can completely be addressed with this mechanism.
